src/workerd/api/gpu/gpu-buffer.c++

```diff
@@ -60,4 +85,83 @@ void GPUBuffer::unmap() {
   }
 }
 
+jsg::Promise<void> GPUBuffer::mapAsync(GPUFlagsConstant mode,
+                                       jsg::Optional<GPUSize64> offset,
+                                       jsg::Optional<GPUSize64> size) {
+  wgpu::MapMode md = static_cast<wgpu::MapMode>(mode);
+
+  if (state_ != State::Unmapped) {
+    device_.InjectError(
+        wgpu::ErrorType::Validation,
+        "mapAsync called on buffer that is not in the unmapped state");
+    JSG_FAIL_REQUIRE(
+        Error, "mapAsync called on buffer that is not in the unmapped state");
+  }
+
+  uint64_t o = 0;
+  KJ_IF_MAYBE (real_offset, offset) {
+    o = *real_offset;
+  }
+
+  uint64_t s = desc_.size - o;
+  KJ_IF_MAYBE (real_size, size) {
+    s = *real_size;
+  }
```

Comment:
Non-blocking: Could possibly use `uint64_t s = size.orDefault(...)` syntax to make this a bit more compact?